Use CPU time for --warn-long

orlitzky opened the pull request:
Finally address #32981.

Using wall time to measure which tests are "long" doesn't make much sense. Wall time is subject to every conflating factor:

1. Faster hardware runs tests in less wall time
2. Faster compiler flags makes code that runs in less wall time
3. A busy machine will run tests slower than an unburdened one
4. Operating system priority can increase or decrease the time taken by a process
5. etc.

Ideally, everyone would agree on how long is too long for a test to run. Attaining that ideal is not so easy, but the first step is to switch`--warn-long` away from using wall time and towards using CPU time. A second phase (https://github.com/sagemath/sage/issues/33022) will normalize the CPU time.


### Examples

#### --warn-long output

The user-facing portion of this can be tested with a low `--warn-long` value,

```
$ sage -t --warn-long=1 src/sage/rings/qqbar.py 
...
Doctesting 1 file.
sage -t --warn-long --random-seed=309742757431591512726158538133919135994 src/sage/rings/qqbar.py
**********************************************************************
File "src/sage/rings/qqbar.py", line 511, in sage.rings.qqbar
Warning, slow doctest:
    z, = [r for r in p4.roots(QQbar, False) if r in ival]
Test ran for 1.51s cpu, 1.51s wall
Check ran for 0.00s cpu, 0.00s wall
**********************************************************************
File "src/sage/rings/qqbar.py", line 524, in sage.rings.qqbar
Warning, slow doctest:
    z2 = QQbar.polynomial_root(p4, ival)
Test ran for 1.07s cpu, 1.07s wall
Check ran for 0.00s cpu, 0.00s wall
**********************************************************************
File "src/sage/rings/qqbar.py", line 808, in sage.rings.qqbar.AlgebraicField_common._factor_multivariate_polynomial
Warning, slow doctest:
    F = QQbar._factor_multivariate_polynomial(p)
Test ran for 2.38s cpu, 2.42s wall
Check ran for 0.00s cpu, 0.00s wall

...
```

#### pexpect accounting

Two different methods are used to account for the CPU time, one for linux, and one for everything else. This GAP command uses pexpect and should chew up a good bit of CPU time:

```
sage: from sage.doctest.util import Timer
sage: t = Timer()
sage: _ = t.start(); t.stop()    # no CPU time used between start/stop
{'cputime': 0.0, 'walltime': 7.605552673339844e-05}

sage: _ = t.start(); gap.eval('a:=List([0..10000],i->WordAlp("a",i));; IsSortedList(a);'); t.stop()    # lots of CPU time used
'true'
{'cputime': 30.089999999999996, 'walltime': 36.292088747024536}
```

orlitzky committed:
src/sage/doctest/control.py: decrease standard --warn-long limit

The default value of 60 cpu seconds for a normal test is too high: the
documentation suggests that a non-long test should complete in about
five seconds. So, in this commit we change the default --warn-long
value to 10 cpu seconds if the --long flag was not given.

Issue: GH-32981

orlitzky committed:
src/sage/doctest/forker.py: report verbose doctest success in CPU time

Since we're now warning about tests that take too many cpu-seconds (as
opposed to too many wall-seconds), it makes sense to report the time
taken in cpu-seconds as well. This allows you to better judge how
close a test is to being a "long" test. The output format remains
unchanged, for example,

  ok [1.02 s]

but now the "s" refers to cpu-seconds instead of wall-seconds. A
subsequent commit will disambiguate the two types of "s".

Issue: GH-32981

orlitzky committed:
src/sage/doctest/forker.py: specify walltime in sage -t --verbose output

Now that cpu seconds are used for --warn-long, we should disambiguate
wherever wall seconds are still used. One such place is in the verbose
test runner output, which outputs (for example)

  ok [3.83 s]

after a test that took 3.83 wall seconds. This commit changes that
output to

  ok [3.83s wall]

which is unmistakable.

Closes: GH-32981

orlitzky committed:
src/sage/doctest/util.py: cputime for pexpect interfaces on Linux/BSD

After switching the doctests to report cputime instead of walltime,
the running-time measurements for tests using pexpect interfaces are
off because the time spent in subprocesses is not accounted
for. Fixing this turns out not to be easy: the cputime() function in
sage.misc.timing can try to account for them, but it is essentially
asking the subprocess to compute its own cputime. This has faults:

  1. Not every pexpect interface implemented (or can implement)
     such a method.

  2. The pexpect interface needs to be in good working condition
     before we can ask the subprocess to do anything. That's often
     not the case at the end of a "sage: ..." line in a doctest.
     Particularly it's not true in all of the tests for low-level
     details of our pexpect interface itself.

Instead, this commit parses /proc/<pid>/stat to obtain the cputime
numbers for pexpext processes. This works well with one caveat: that
information is only available on Linux and BSD.

Having a solution that only works on some platforms is not actually
too problematic. The motivation for measuring these times in the
doctest framework is that someone should notice an outrageously long
test and fix or report it. So, it's enough that the measurements be
accurate on only some platforms -- particularly where the CI is
run. Furthermore, the information is only both missing and potentially
useful for long-running pexpect tests, which are relatively rare.

It would be possible to implement the same thing for macOS using its
own platform-specific interface if anyone is motivated to do so. It
would also make sense to use psutil to obtain this information once
we have upgraded to a version of ipykernel that reintroduces it as
a dependency (see GH-33772).

Issue: GH-32981

orlitzky committed:
src/sage/doctest/forker.py: report both cpu/wall times for long tests

We need to report the CPU time for tests that run afoul of
--warn-long, because --warn-long is now specified in CPU time. But, we
are still reporting the wall time in several places, such as the
per-file summary. So to keep things as consistent as possible, we
should report the wall time for --warn-long failures as well. This
changes the output of a too-long test from e.g.

  Test ran for 1.23 cpu seconds, check ran for 2.34 cpu seconds

to,

  Test ran for 1.23s cpu, 2.50s wall
  Check ran for 2.34s cpu, 3.12s wall

Issue: GH-32981

orlitzky committed:
src/sage/doctest: specify walltime in per-file doctest report

Now that cpu seconds are used for --warn-long, we should disambiguate
wherever wall seconds are still used. One such place is in the doctest
runner's per-file report, which outputs (for example)

  [14 tests, 0.26 s]

after testing a file with 14 tests that took 0.26 wall seconds. This
commit changes that output to

    [14 tests, 0.26s wall]

which is explicit about the type of measurement used.

Issue: GH-32981

orlitzky committed:
src/sage/doctest: use CPU time for --warn-long

Our --warn-long flag used during testing currently takes "wall time"
as its argument. This is somewhat nice from a user perspective, since
a normal user is quite likely to know what what wall time is, but it's
also a bit nonsensical given the purpose of the flag. The flag's
intent is to warn developers about tests that take too long, so
whether or not it triggers should not depend on the developer's
machine, or what else he's doing while the tests are running -- both
of which do affect the wall time.

CPU time provides a more reliable measurement. (In fact, when no value
for --wall-time was given, we were already using a heuristic to mimic
CPU time.) This commit changes the --warn-long flag to use CPU time
instead of wall time, obviating the need for the heuristic.

Issue: GH-32981

orlitzky committed:
src/doc/en/developer/doctesting.rst: we use CPU time now

Update the documentation to reflect that we are now using CPU time
(and not wall time) to measure the length of a test.

Issue: GH-32981

orlitzky committed:
src/bin/sage{,-runtests}: mention CPU time in --warn-long

The --warn-long flag now takes CPU seconds (not wall seconds) as its
argument. Here we update some command-line output to be explicit about
this fact.

Issue: GH-32981

tornaria commented on src/sage/doctest/util.py:
```diff
@@ -103,6 +103,68 @@ class Timer:
+                try:
+                    cputime += self._pid_cpu_seconds(S.pid())
+                except (ArithmeticError, LookupError, OSError,
+                        TypeError, ValueError):
```
Maybe it's better to catch errors in `_pid_cpu_seconds` and raise a well specified exception there?

Also: a. why are these methods? and b. why not fix `cputime(...,subprocesses=True)` instead?

orlitzky replied:
> Maybe it's better to catch errors in `_pid_cpu_seconds` and raise a well specified exception there?

I can't remember why I did it this way in the first place, but my best guess is that raising the exception in `_pid_cpu_seconds` would involve a little bit more code for no added benefit. Instead of,

```python
except (ArithmeticError, LookupError, OSError,
                        TypeError, ValueError):
```

we'd have both,

```python
except (ArithmeticError, LookupError, OSError,
                        TypeError, ValueError):
    raise OSError
...
except OSError:
    ...
```

> Also: a. why are these methods?

Just for namespacing. They're not meant to be used outside of the `Timer` class, and the leading underscore makes them pseudo-private by python convention.

> and b. why not fix `cputime(...,subprocesses=True)` instead?

The new method does something fundamentally different, and doesn't always work. For the purposes of `--warn-long`, it's fine if we mis-account for a few CPU seconds on macOS, but I wouldn't want to have to justify breaking `cputime(subprocesses=True)` on macOS in whatever situations it happens to work right now.

orlitzky replied:
> Maybe it's better to catch errors in `_pid_cpu_seconds` and raise a well specified exception there?

I thought about this for a while and I think it's a better design. It will take me all day to rebuild sage but I've pushed the commit so if I'm real lucky, the CI will pass. In the process I documented where all of the various exceptions can come from.

tornaria commented on src/sage/doctest/control.py:
```diff
@@ -600,7 +601,12 @@ def _init_warn_long(self):
+        # being rather generous here.
+        self.options.warn_long = 10.0
+        if self.options.long:
+            self.options.warn_long = 60.0
```
Could we aim for 5.0 and 30.0? That's already 5x margin from what is recommended ( < 1s for normal tests / < 5s for long tests).

Are we too far from that? I don't think there are currently more than 50-100 tests that fail that and then it shouldn't be too hard to "fix" those (or just add `# long` or `# not tested` labels).

orlitzky replied:
Sure, I was trying not to ruffle any feathers, but I think 5s and 30s is still very reasonably. Done in a new commit. Ultimately it would be nice if we could achieve something like 1s/5s.

orlitzky committed:
src/sage/doctest/control.py: further reduce the --warn-long limit

Limits of 5s and 30s for normal and long tests, respectively, are more
in line with the suggestions in the developer's guide (which are more
strict still). So, we cut the existing limits in half to achieve that.

orlitzky committed:
src/sage/doctest/util.py: update a comment

Replace "sage.misc.misc.cputime" with "sage.misc.timing.cputime",
because it moved.

orlitzky committed:
src/sage/doctest/util.py: refactor exception handling

Instead of handling all of the possible cputime exceptions in the
Timer class's _quick_cputime(), we move it into _pid_cpu_seconds(),
closer to where the errors actually occur. This allows us to re-raise
an OSError in all exceptional cases, making them easier to handle
higher up. In the process some of the exceptions' causes have been
documented.

jhpalmieri commented on src/sage/doctest/control.py:
```diff
@@ -575,36 +575,38 @@ def __del__(self):
-        The default used here is 60 seconds on a modern computer. It
-        should eventually be lowered to 5 seconds, but its best to
-        boil the frog slowly.
+        The default used here is 10 seconds, unless `--long` was used,
```
``
```suggestion
        The default used here is 5 seconds, unless `--long` was used,
```

jhpalmieri replied:
(for consistency with the actual code)

orlitzky replied:
Thanks, fixed.

jhpalmieri commented on src/sage/doctest/control.py:
```diff
@@ -575,36 +575,38 @@ def __del__(self):
-        should eventually be lowered to 5 seconds, but its best to
-        boil the frog slowly.
+        The default used here is 10 seconds, unless `--long` was used,
+        in which case it is 60 seconds.
```
```suggestion
        in which case it is 30 seconds.
```

orlitzky replied:
Also fixed.

orlitzky committed:
src/sage/doctest/control.py: update _init_warn_long() docstring

Now that the --warn-long defaults are 5.0 and 30.0 seconds for regular
and long tests respectively, we update the _init_warn_long() docstring
to reflect that.

jhpalmieri commented on src/sage/doctest/util.py:
```diff
@@ -103,6 +103,95 @@ class Timer:
         sage: TestSuite(Timer()).run()
     """
+
+    def _pid_cpu_seconds(self, pid):
```
Can you produce some doctests for this?

orlitzky replied:
These functions are basically random number generators, but I've done what I can.

jhpalmieri commented on src/sage/doctest/forker.py:
```diff
@@ -1467,7 +1475,7 @@ def report_failure(self, out, test, example, got, globs):
             return returnval
 
-    def report_overtime(self, out, test, example, got, *, check_duration=0):
+    def report_overtime(self, out, test, example, got, *, check_timer=None):
```
The old keyword should be deprecated using `sage.misc.decorators.rename_keyword`, rather than just changed.

orlitzky replied:
The keyword wasn't simply renamed -- the type of argument that it takes was changed too. Typically that would be an unfriendly thing to do, but this is essentially a private method. No other part of the sage library uses it (it's referenced exactly once, by `self`), and since it's part of sage's custom doctest runner, it's hard to imagine a third party relying on these internal details. Frankly I'd like to change it to an underscore method now that you mention it.

jhpalmieri replied:
> The keyword wasn't simply renamed -- the type of argument that it takes was changed too. Typically that would be an unfriendly thing to do, but this is essentially a private method. No other part of the sage library uses it (it's referenced exactly once, by `self`), and since it's part of sage's custom doctest runner, it's hard to imagine a third party relying on these internal details. 

Cool. Sage policy is still to deprecate it. So deprecate the old argument and add a new one.

orlitzky replied:
There's no way to translate the old argument into the new one. If we keep the old name but ignore its argument, then any code using this function will still be immediately broken, only more quietly and with more steps.

orlitzky replied:
If you feel that strongly about not changing this method's signature, I can copy & paste the entire function into a new private method `_report_overtime`. And then we should do the same for `_report_success`. That will leave the old methods intact to be deprecated, but I really think we would be wasting our time.

It's going to take time to put the old functions back and to add the deprecation warnings. It will take time to re-review those changes. It's going to take time to run the doctests for two unused functions in the interim. And then eventually someone has to spend the time to remove the functions, and someone else has to spend the time to review that.

Unless there's an (even remotely) plausible reason why anyone would be calling this function in their own code, it's justifiable to bend the deprecation rule. Its intent has been respected. And to do otherwise is wasting our most valuable resource.

jhpalmieri commented on src/sage/doctest/util.py:
```diff
@@ -103,6 +103,95 @@ class Timer:
+
+        return (cputicks / hertz)
+
+    def _quick_cputime(self):
```
Doctests?

orlitzky replied:
In case my other comment didn't make it clear, I added a doctest for this function too, though it isn't much.

orlitzky committed:
src/sage/doctest/util.py: delete a wandering "and"

Remove an errant "and" from the docstring for Timer._quick_cputime().

orlitzky committed:
src/sage/doctest/util.py: add doctests for Timer._quick_cputime()

The output of this function is "random", but we can at least check
that the returned value is a nonnegative float.

orlitzky committed:
src/sage/doctest/util.py: improve docstring for Timer._pid_cpu_seconds()

Add INPUT, OUTPUT, and TESTS blocks to the _pid_cpu_seconds() method
of the Timer class.

orlitzky committed:
src/sage/doctest/util.py: reword docstring to make flake8 happy

orlitzky committed:
src/sage/doctest/control.py: update the output from one more test

orlitzky committed:
src/sage/doctest/reporting.py: spaces around % for pycodestyle

orlitzky committed:
src/sage/doctest/util.py: fix _pid_cpu_seconds doctest on macOS

In doctests, cputime = 0.0 gets preparsed into a RealLiteral causing
isinstance(cputime, float) to return False.

orlitzky committed:
pkgs/sagemath-repl/MANIFEST.in: include sage.interfaces.{cleaner,quit}

These two modules are now needed by the doctest runner to report the
CPU time used by each test.

orlitzky committed:
sage/interfaces/{cleaner,quit}.py: tag as distribution = sagemath-repl

These are now included in the sagemath-repl distribution because the
doctest runner uses them.

orlitzky committed:
pkgs/sagemath-repl/pyproject.toml.m4: add package sage.interfaces

The sagemath-repl distribution now ships two modules belonging to
sage.interfaces. If we don't include sage.interfaces in the "packages"
list, they won't actually be installed.

orlitzky committed:
src/sage/doctest/util.py: add error messages and test one of them

The errors from _pid_cpu_seconds() aren't actually visible to end
users because _quick_cputime() eats them, but for the sake of code
coverage, we force one to happen with an invalid PID and test the
result.

orlitzky committed:
src/sage/doctest/util.py: improve a comment

The /proc/<pid>/stat fields could use a bit more documentation. And it
looks like they've moved from a man page to the kernel source tree.

orlitzky committed:
src/sage/doctest/util.py: refactor _quick_cputime() for testability

Instead of importing the global list of expect_objects from inside the
_quick_cputime() function, it now takes them as a argument. This
allows us to test what happens when some pexpect process causes an
error further down in _pid_cpu_seconds().

orlitzky committed:
src/sage/doctest/util.py: refactor for testability

The _pid_cpu_seconds(self, pid) function was hard to test because no
matter what you pass in, it tries to read from /proc. It has been
renamed to _proc_stat_cpu_seconds(self, path), which now takes a path
instead. This will make it easier to test the error paths by passing
in paths to carefully constructed junk files.

orlitzky committed:
src/sage/doctest/util.py: test _proc_stat_cpu_seconds() error paths

Now that _proc_stat_cpu_seconds() takes a path instead of a PID, we
add two new tests for error conditions. The first checks for too few
fields in the stat file, and the second checks for non-float
fields. Both use a temporary file whose path can now be passed in.

orlitzky committed:
src/sage/doctest/util.py: support macOS if psutil is installed

If psutil is installed (it usually will be, because ipython uses it),
we can use it to compute the cputime used by pexpect processes on
macOS. This is of course guarded by a try/except clause, and may not
always work, but is better than giving up immediately on macOS.

orlitzky committed:
src/sage/doctest/util.py: ignore ZombieProcess errors from psutil

Apparently psutil will raise a custom ZombieProcess exception if you
try to compute the cputime of a zombie process. No problem, we can
ignore those too.

orlitzky committed:
src/sage/doctest/util.py: avoid dependency on sage.misc.timing.walltime

The "walltime" function in sage.misc.timing is a silly wrapper around
time.time() in the python standard library. The code is a lot more
clear if we just use the latter.

orlitzky committed:
src/sage/doctest/forker.py: remove old python 2 workaround

According to the comment, we can skip clearing this exception nowadays
because we only support python3.

orlitzky committed:
src/sage/doctest/util.py: handle NoSuchProcess from psutil, too

Like ZombieProcess... this can happen. And the doctest framework
shouldn't crash if a pexpect process was killed.

orlitzky committed:
src/sage/doctest/util.py: put "import" inside the "try"

We're importing psutil and trying to catch the ImportError if one is
raised. But for that to work, it helps to put "import" inside of the
"try" block.

orlitzky committed:
src/sage/doctest/forker.py: re-raise SignalErrors

If SystemExit occurs during a doctest, we re-raise it. The same thing
should be done for SignalError (from cysignals). Signals can occur at
any time, and the doctest runner shouldn't eat them.

In particular, this can lead to a Heisenbug where a signal will
terminate a doctest example before the timer has a chance to annotate
the it with the elapsed walltime and cputime. In that case, we get
an "impossible" AttributeError later on when we try to compute the
total walltime used by the example.

tobiasdiez commented on src/sage/doctest/control.py:
```diff
@@ -588,36 +588,38 @@ def __del__(self):
+        if self.options.long:
+            self.options.warn_long = 30.0
 
     def second_on_modern_computer(self):
```
This method can now be deleted, right?

orlitzky replied:
Yes, but it's a "public" function, so I didn't want to start an argument over the deprecation policy.

tobiasdiez replied:
I think it is safe to consider everything in doctest as private. But as a compromise, at least add a deprecation warning so that this can be removed in 1year?

orlitzky replied:
Done

tobiasdiez reviewed (commented):
The tests report slow docstests although those are marked as "long time" (see annotations at the bottom in https://github.com/sagemath/sage/actions/runs/11373830486?pr=36226). Is this by design because those tests take longer than 30s? If yes, what is the plan with these "super long" tests?

Other than this point, the PR looks good to me.

orlitzky committed:
sage/doctest/control.py: deprecate second_on_modern_computer()

This method of the DocTestController class is no longer used in the
sage library, but technically, it was "public," so we deprecate it for
later removal.

orlitzky committed:
src/sage/doctest/forker.py: change the slow doctest warning punctuation

Switch "Warning, slow doctest:" to "Warning: slow doctest:" so that
the leading "Warning: " is consistent with other warnings. This
ultimately prevents these slow doctest warnings from being converted
into error annotations on the Github CI.

tobiasdiez reviewed (approved):
I would prefer if they wouldn't show up at all, but it's fine like this for me. 

Feel free to set it to positive review, or wait a little bit if you still want a second pair of eyes to look at the code.
